### PR TITLE
Improve sorting for strings and push projections into sort operator

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR improves sorting for strings by using pdqsort instead of radix sort when there are strings. Radix sort is less effective on strings because the fan-out is generally very low, so pdqsort is slightly faster, especially when the strings are long.

I've also added the `ProjectionCombiner` optimizer, that pushes projections into the `PhysicalOrder` operator, and can be extended to push projection into other operators as well. Previously, when running a query like this:
```sql
SELECT c_customer_sk
FROM customer
ORDER BY c_last_name, c_first_name;
```
We would get a plan that duplicated the key columns, sorted them, and then projected them out later. The optimizer pushes the projection into the operator, and now we no longer have duplicated columns.

I've run the following query on TPC-DS SF300 on an AWS `m5d.8xlarge` box:
```sql
SELECT count(c_customer_sk)
FROM (
    SELECT c_customer_sk
    FROM customer
    ORDER BY c_last_name, c_first_name
);
-- This query sorts one integer column by two string columns
```

With these new optimizations the query completes in around 0.195 seconds. Before, it completed in around 0.394. The biggest difference is the de-duplication of the columns. The switch to PDQSort only yields about a 5-10% improvement.

### CODE REVIEW

File: src/include/duckdb/execution/operator/order/physical_order.hpp [hunk 1]
```diff
@@ -20,10 +20,12 @@ class OrderGlobalState;
 //! Physically re-orders the input data
 class PhysicalOrder : public PhysicalOperator {
 public:
-	PhysicalOrder(vector<LogicalType> types, vector<BoundOrderByNode> orders, idx_t estimated_cardinality);
+	PhysicalOrder(vector<LogicalType> types, vector<BoundOrderByNode> orders,
+	              vector<unique_ptr<Expression>> projections, idx_t estimated_cardinality);
 
 	//! Input data
 	vector<BoundOrderByNode> orders;
+	vector<unique_ptr<Expression>> projections;
```
Comment (Mytherin) [inline src/include/duckdb/execution/operator/order/physical_order.hpp:28]: I'm not quite convinced that it is necessary to add projection columns to the order node itself. Can't we get the same functionality by pushing a projection node before the order in the plan?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. One comment/question:

File: src/include/duckdb/execution/operator/order/physical_order.hpp (hunk 1 above)
Comment (lnkuiper) [reply src/include/duckdb/execution/operator/order/physical_order.hpp:28]: Let's say we have this query:
```sql
SELECT a
FROM tbl
ORDER BY a, b, c;
```

This plan is generated:
```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             #0            │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          ORDER_BY         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         tbl.a ASC         │
│           b ASC           │
│           c ASC           │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│             c             │
└───────────────────────────┘                   
```

We need columns `a`, `b`, and `c` in the `ORDER_BY` operator, so we can't project any columns out beforehand, so we project them out afterwards. However, this causes the `ORDER_BY` operator to duplicate the `a`, `b`, and `c` columns in both the key and payload columns within the operator.

By pushing the projection into the operator, we will only add `a` to the payload columns, meaning less column duplication.

This optimizer can also be used to push projections into a table scan, which is useful when the table scan is performed by a remote instance. For example, this query:
```sql
EXPLAIN SELECT b, c FROM tbl WHERE a > 42;
```

We get this plan:
```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             b             │
│             c             │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            tbl            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             a             │
│             b             │
│             c             │
└───────────────────────────┘      
```

While the `SEQ_SCAN` is performed by a remote DuckDB instance, the `PROJECTION` may be performed locally. However, at this point the `a` column travels over the network, even though it is immediately projected out. If we push the projection into the table scan we can prevent this.

**Comment by Mytherin** [comment]
No, that R test failure is a sporadic failure that has been happening for a while.

**Review by Mytherin — APPROVED** [review]
Thanks for the changes! LGTM now

**Comment by lnkuiper** [comment]
Ah I'm sorry, I shouldn't have pushed this yet. I've added the optimization for this query:
```sql
SELECT b, c FROM tbl WHERE a > 42;
```
Where `a` does not leave the table scan at all.

If CI passes though, I think this optimization can go in too.

**Comment by Alex-Monahan** [comment]
Hey @lnkuiper! Is it worth doing a small new blog as an update to the sorting blog post? I know you have significantly sped up sorting in a few ways since that post!

**Pull request merged**